src/mdev.rs

```diff
@@ -0,0 +1,472 @@
+//! Structures for representing a mediated device
+
+use crate::environment::Environment;
+use anyhow::{anyhow, Context, Result};
+use log::{debug, warn};
+use std::convert::TryInto;
+use std::fs;
+use std::path::{Path, PathBuf};
+use std::vec::Vec;
+use uuid::Uuid;
+
+#[derive(Clone, Copy)]
+pub enum FormatType {
+    Active,
+    Defined,
+}
+
+/// Representation of a mediated device
+#[derive(Debug)]
+pub struct MDev<'a> {
+    pub uuid: Uuid,
+    pub active: bool,
+    pub autostart: bool,
+    pub parent: Option<String>,
+    pub mdev_type: Option<String>,
+    pub attrs: Vec<(String, String)>,
+    env: &'a dyn Environment,
+}
+
+impl<'a> MDev<'a> {
+    pub fn new(env: &'a dyn Environment, uuid: Uuid) -> MDev<'a> {
+        MDev {
+            uuid,
+            active: false,
+            autostart: false,
+            parent: None,
+            mdev_type: None,
+            attrs: Vec::new(),
+            env,
+        }
+    }
+
+    pub fn path(&self) -> PathBuf {
+        let mut p = self.env.mdev_base();
+        p.push(self.uuid.to_hyphenated().to_string());
+        p
+    }
+
+    // get parent and propagate a consistent error to the caller if absent
+    pub fn parent(&self) -> Result<&String> {
```

Comment:
Could this return `&str` instead? Is there a specific reason you're returning a borrowed owned type (String) vs just a borrowed type (&str)?

Reply:
Something like this? 

```suggestion
    pub fn parent(&self) -> Result<&str> {
                self.parent.map(|x| x.as_str()).ok_or_else(|| {
```
(replaced as_ref() with map())

It's not clear to me what we would gain from a change like this.